repl/src/main/scala/ammonite/repl/interp/Interpreter.scala

```diff
@@ -311,12 +328,70 @@ class Interpreter(prompt0: Ref[String],
       (classFiles, newImports) <- compiled
       _ = Timer("cachedCompileBlock 4")
       cls <- eval.loadClass(fullyQualifiedName, classFiles)
-    } yield (cls, newImports)
+    } yield (cls, newImports, tag)
+  }
+
+
+  def evalCachedClassFiles(cachedData: Seq[ClassFiles],
+                           pkg: String,
+                           wrapper: String): Res[Seq[Unit]] = {
+
+    def getBlockNumber(blockNumber: Int) = blockNumber match {
+      case 1 => ""
+      case _ => "_" + blockNumber.toString
+    }
+
+      val cls = Res.map(cachedData.zipWithIndex)(
+        (data: (ClassFiles, Int)) => {
+          val (clsFiles, index) = data
+          Compiler.addToClasspath(clsFiles, dynamicClasspath)
+          eval.loadClass(pkg + "." + wrapper + getBlockNumber(index + 1), clsFiles)
+        }
+      )
+      cls.map(_.map(eval.evalMain(_)))
```

Comment:
You should use a `for`-comprehension instead of `.map` on the `Res`. 

The name `cls` is also misleading, since this isn't a `Class`, it is a `res`.

Furthermore, this is incorrect, as an exception inside `evalMain` will get thrown and not caught. You need to use a `_ <- Catching{userCodeExceptionHandler}` inside your for-comprehension to catch that.

Lastly, this stuff which has to do with evaluating compiled class files should live in `Evaluator`, not in `Interpreter`